### PR TITLE
Add architecture documentation

### PR DESCRIPTION
**Pull request by danielballan**
Credit to @padraic-shafer for prompting this much needed addition

### CODE REVIEW

**Comment by padraic-shafer** [comment]
Wow, I didn't even have to do any work and I get credit? 🤫

File: docs/source/explanations/architecture.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+## Client
+
+Tiled ships with a Python client. This is separable from the rest of the
+project, and could potentially someday be split off as a separate package. The
+client enables the user to navigate and access the data in a Tiled server using
+familiar Python item-lookup and slicing idioms, while it generates HTTP
+requests to transfer metadata and data.
+
+When the user connects, a **Context** object is created. The Context is
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:13]:
```suggestion

### Connection _Context_
When the user connects, a **Context** object is created. The Context is
```

File: docs/source/explanations/architecture.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+shared by all client-side objects that use this connection. It wraps an
+[httpx HTTP client][httpx], which in turn wraps an HTTP connection pool and (if
+applicable) authentication-related state---either an API key or a pair of
+OAuth2 access and refresh tokens.
+
+The Context also may hold an HTTP response **Cache**, similar to a web
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:19]:
```suggestion

### Client-side _Caching_
The Context also may hold an HTTP response **Cache**, similar to a web
```

File: docs/source/explanations/architecture.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+Most endpoints require authentication, unless the server is configured to be
+public. For single-user deployments, a single API key is specified or randomly
+generated at server startup. For multi-user deployments, an **Authentication
+Database** (PostgreSQL or SQLite) is used to store session information and to
+validate API keys.
+
+Endpoints that serve metadata or data resolve the URL path to identify the
+relevant **Adapter** (about which, more below), which returns the data
+as a scientific data structure. It may be a "lazy" data structure,
+representing data that will be loaded later, on demand, piecemeal.
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:39]:
```suggestion
representing data that will be loaded later -- on demand and piecemeal.
```

File: docs/source/explanations/architecture.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+experimental.
+
+## Server
+
+The Tiled HTTP server is built using the framework [FastAPI][], which is built
+on [Starlette][]. A key feature of FastAPI is auto-generated [OpenAPI][]
+documentation, which Tiled serves at the `GET /docs` endpoint. FastAPI
+works with [Pydantic][] to parse and validate requests.
+
+Most endpoints require authentication, unless the server is configured to be
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:30]:
```suggestion

### Authentication
Most endpoints require authentication, unless the server is configured to be
```

File: docs/source/explanations/architecture.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+public. For single-user deployments, a single API key is specified or randomly
+generated at server startup. For multi-user deployments, an **Authentication
+Database** (PostgreSQL or SQLite) is used to store session information and to
+validate API keys.
+
+Endpoints that serve metadata or data resolve the URL path to identify the
+relevant **Adapter** (about which, more below), which returns the data
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:37]:
```suggestion
relevant [**Adapter**](#adapter), which returns the data
```

File: docs/source/explanations/architecture.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+as a scientific data structure. It may be a "lazy" data structure,
+representing data that will be loaded later, on demand, piecemeal.
+
+The endpoint implements [content negotiation][], comparing the list of formats
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:41]:
```suggestion

### Content Negotiation and _Serializers_
The endpoint implements [content negotiation][], comparing the list of requested formats
```

File: docs/source/explanations/architecture.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+and accepted by the client to those supported by the server for this particular
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:42]:
```suggestion
that are accepted by the client to those supported by the server for this particular
```

File: docs/source/explanations/architecture.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+dataset. It dispatches to a registry of **Serializers**, which convert the data
+structure into bytes which can be sent in a response by FastAPI. Custom
+serializers may be registered in server configuration.
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:45]:
```suggestion
serializers may be registered during server configuration.
```

File: docs/source/explanations/architecture.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+documentation, which Tiled serves at the `GET /docs` endpoint. FastAPI
+works with [Pydantic][] to parse and validate requests.
+
+Most endpoints require authentication, unless the server is configured to be
+public. For single-user deployments, a single API key is specified or randomly
+generated at server startup. For multi-user deployments, an **Authentication
+Database** (PostgreSQL or SQLite) is used to store session information and to
+validate API keys.
+
+Endpoints that serve metadata or data resolve the URL path to identify the
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:36]:
```suggestion

### Accessing Data and Metadata
Endpoints that serve metadata or data resolve the URL path to identify the
```

File: docs/source/explanations/architecture.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+as a scientific data structure. It may be a "lazy" data structure,
+representing data that will be loaded later, on demand, piecemeal.
+
+The endpoint implements [content negotiation][], comparing the list of formats
+and accepted by the client to those supported by the server for this particular
+dataset. It dispatches to a registry of **Serializers**, which convert the data
+structure into bytes which can be sent in a response by FastAPI. Custom
+serializers may be registered in server configuration.
+
+On its way, the response may be compressed, again using content negotiation
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:47]:
```suggestion

### Compression
On its way, the response may be compressed, again using content negotiation
```

File: docs/source/explanations/architecture.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+In Tiled, an **Adapter** provides a standard interface to data, regardless of
+how it is stored. Adapters that wrap different structure have different
+interfaces. For example, an array adapter implements `read_block` whereas a
+table adapter implements `read_partition`. But all array adapters are alike,
+and all table adapters are alike. They enable the server to abstract over
+the details of _how_ to get the data.
+
+Typically Adapters take a filepath or URI and, perhaps, some additional
+configuration. For development, test, or demo purposes, Adapters can wrap data
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:61]:
```suggestion
configuration. For development, test, or demonstration purposes, Adapters can wrap
```

File: docs/source/explanations/architecture.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+in memory, like a numpy array. Several Adapters are included in the Tiled
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:62]:
```suggestion
in-memory data, such as a numpy array. Several Adapters are included in the Tiled
```

File: docs/source/explanations/architecture.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+codebase, in the spirit of "batteries included," but Adapters can be defined in
+external modules, too, and operate on the same footing as the built-in ones.
+
+## Catalog
+
+The Catalog is an Adapter that stores the metadata and structure for a
+potentially large number of datasets in a SQL database (PostgreSQL or SQLite).
+This enables it to efficiently response to _metadata_ requests and _search_
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:70]:
```suggestion
This enables it to efficiently respond to _metadata_ requests and _search_
```

File: docs/source/explanations/architecture.md [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+requests without opening any data files. Requests for _data_ are then
+dispatched down to the appropriate Adapter which can load the data
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:72]:
```suggestion
dispatched down to the appropriate [Adapter](#Adapter) which can load the data
```

File: docs/source/explanations/architecture.md [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+But for most standard applications, including serving a directory of files or
+providing a writable data store, the Catalog is used.
+
+[FastAPI]: https://fastapi.tiangolo.com/
+[httpx]: https://www.python-httpx.org/
+[Starlette]: https://www.starlette.io/
+[OpenAPI]: https://www.openapis.org/
+[Pydantic]: https://docs.pydantic.dev/
+[content negotiation]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation
```
Comment (padraic-shafer) [inline docs/source/explanations/architecture.md:90]: This is a nice paradigm for tidying links in the docs.  I haven't seen it before.

**Review by padraic-shafer — APPROVED** [review]
Very nice!

I think that adding subheadings will help draw the reader's eye to key topics.  Also they can be linked to in other docs & discussions.

**Comment by danielballan** [comment]
Good edits. I'm stuck on getting any of the `#adapter` references to work; the build fails on:

Warning, treated as error:
/home/dallan/Repos/bnl/tiled/docs/source/explanations/architecture.md:40:'myst' cross-reference target not found: 'adapter' [myst.xref_missing]
```

**Comment by padraic-shafer** [comment]
For the links to <H2> tags we might need to set `myst_heading_anchors=2` in the sphinx [build options](https://github.com/bluesky/tiled/blob/main/docs/Makefile).  I haven't had to do this before, but it's referenced [here](https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#implicit-targets).

**Comment by danielballan** [comment]
I hope we can figure out how to make explicit targets work. I agree with this advice in the docs:

> In general, it is discouraged to rely on implicit targets, since they are easy to break, if for example a document/heading is moved or renamed.

**Comment by padraic-shafer** [comment]
> I hope we can figure out how to make explicit targets work.

I'm trying it locally now.  Will report back.

**Comment by danielballan** [comment]
Ah, I thought that the target was an arbitrary string. I did not realize that the suffix `-target` was significant.

**Comment by danielballan** [comment]
Oh, you're right. Then I'm not sure what I was doing wrong in some local experiments earlier. Anyway, glad we have a working pattern now.

**Review by dylanmcreynolds — COMMENTED** [review]
Would this be a good place to say (or link to) install client vs server via pip?

**Comment by padraic-shafer** [comment]
Similarly, the end of each sub-section might contain links to the relevant code documentation and further info.

_E.g._
* https://blueskyproject.io/tiled/reference/python-client.html
* 
* https://blueskyproject.io/tiled/explanations/caching.html
    * https://blueskyproject.io/tiled/reference/python-client.html#cache
* 
* https://blueskyproject.io/tiled/how-to/tiled-authn-database.html
    * https://blueskyproject.io/tiled/reference/service-configuration.html#authentication
    * https://blueskyproject.io/tiled/tutorials/login.html
* 
* https://blueskyproject.io/tiled/reference/service.html#adapters
* 
* https://blueskyproject.io/tiled/reference/service.html#media-type-format-registry

**Review by padraic-shafer — APPROVED** [review]
Built locally; all ref links are working correctly.

**Pull request merged**